# Data preparation - Mushrooms project



### Import used modules

The only external modules used for this notebook are [numpy](http://www.numpy.org/), and [pandas](https://pandas.pydata.org/).

In [22]:
from collections import Counter  
# it is an unordered collection where elements and their respective count are stored as a dictionary
#  print(Counter(['B','B','A','B','C','A','B','B','A','C']))
# o/p Counter({'B': 5, 'A': 3, 'C': 2})

from pathlib import Path

import pandas as pd
import numpy as np

## Explore the data

The first step is to explore the data, so we can see what we are dealing with. We will use the pandas module to load the data from the .csv file.

In [23]:
data = pd.read_csv("data/mushrooms.csv")
data

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

In [24]:
data.shape

(8124, 23)

The data consist of 8124 different inputs with 22 features and one label (the **"class"** column). 
We can see that all the features (colums) are categorical. This means that that feature can take one of different values and those are not numerical. This means that a neural network can't process them as they are, we will need to convert them to numbers, so the model can perform mathematical operations on them.

We will now check the posible values each feature can take:

In [25]:
possible_values = {}
for feature in data.columns:
    counter = Counter(list(data[feature]))
    print(counter)
    possible_values[feature] = list(counter.keys())
possible_values

Counter({'e': 4208, 'p': 3916})
Counter({'x': 3656, 'f': 3152, 'k': 828, 'b': 452, 's': 32, 'c': 4})
Counter({'y': 3244, 's': 2556, 'f': 2320, 'g': 4})
Counter({'n': 2284, 'g': 1840, 'e': 1500, 'y': 1072, 'w': 1040, 'b': 168, 'p': 144, 'c': 44, 'u': 16, 'r': 16})
Counter({'f': 4748, 't': 3376})
Counter({'n': 3528, 'f': 2160, 'y': 576, 's': 576, 'a': 400, 'l': 400, 'p': 256, 'c': 192, 'm': 36})
Counter({'f': 7914, 'a': 210})
Counter({'c': 6812, 'w': 1312})
Counter({'b': 5612, 'n': 2512})
Counter({'b': 1728, 'p': 1492, 'w': 1202, 'n': 1048, 'g': 752, 'h': 732, 'u': 492, 'k': 408, 'e': 96, 'y': 86, 'o': 64, 'r': 24})
Counter({'t': 4608, 'e': 3516})
Counter({'b': 3776, '?': 2480, 'e': 1120, 'c': 556, 'r': 192})
Counter({'s': 5176, 'k': 2372, 'f': 552, 'y': 24})
Counter({'s': 4936, 'k': 2304, 'f': 600, 'y': 284})
Counter({'w': 4464, 'p': 1872, 'g': 576, 'n': 448, 'b': 432, 'o': 192, 'e': 96, 'c': 36, 'y': 8})
Counter({'w': 4384, 'p': 1872, 'g': 576, 'n': 512, 'b': 432, 'o': 192, 'e': 96, 'c

{'class': ['p', 'e'],
 'cap-shape': ['x', 'b', 's', 'f', 'k', 'c'],
 'cap-surface': ['s', 'y', 'f', 'g'],
 'cap-color': ['n', 'y', 'w', 'g', 'e', 'p', 'b', 'u', 'c', 'r'],
 'bruises': ['t', 'f'],
 'odor': ['p', 'a', 'l', 'n', 'f', 'c', 'y', 's', 'm'],
 'gill-attachment': ['f', 'a'],
 'gill-spacing': ['c', 'w'],
 'gill-size': ['n', 'b'],
 'gill-color': ['k', 'n', 'g', 'p', 'w', 'h', 'u', 'e', 'b', 'r', 'y', 'o'],
 'stalk-shape': ['e', 't'],
 'stalk-root': ['e', 'c', 'b', 'r', '?'],
 'stalk-surface-above-ring': ['s', 'f', 'k', 'y'],
 'stalk-surface-below-ring': ['s', 'f', 'y', 'k'],
 'stalk-color-above-ring': ['w', 'g', 'p', 'n', 'b', 'e', 'o', 'c', 'y'],
 'stalk-color-below-ring': ['w', 'p', 'g', 'b', 'n', 'e', 'y', 'o', 'c'],
 'veil-type': ['p'],
 'veil-color': ['w', 'n', 'o', 'y'],
 'ring-number': ['o', 't', 'n'],
 'ring-type': ['p', 'e', 'l', 'f', 'n'],
 'spore-print-color': ['k', 'n', 'u', 'h', 'w', 'r', 'o', 'y', 'b'],
 'population': ['s', 'n', 'a', 'v', 'y', 'c'],
 'habitat': ['u'

## Labels

First, we will extract the labels from the main data and convert the `p` (poisonous) and `e` (edible) tags to numbers. We will use: `p = 0` and `e = 1`.

In [26]:
# Create a list with all the data points labels
labels_data = list(data["class"]) # taking one column and creating lables
print(labels_data)
# Convert to numbers
labels = []
for label in labels_data:
    if label == "p":
        labels.append([0])
    elif label == "e":
        labels.append([1])
        
labels = np.array(labels)
print("Labels array shape = {}".format(labels.shape))

# Delete "class" column from the data
data = data.drop(labels=["class"], axis=1)

['p', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'p', 'p', 'p', 'e', 'p', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'p', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'p', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e', 'e',

## Features

There are different aproaches to how to handle the categorical data of the features. One way will be to convert each label for each feature into a different number. In example, if we look at the feature column `cap-surface`, it can take the tags `['s', 'y', 'f', 'g']`. We could convert them to `[0, 1, 2, 3]`, but when we use this values on the inputs of the model, the class `g` of `cap-shape` will have the value `3`, and the class `y` will be `1`. For the model this are numerical values, so the incidence of class `g` will be 3 times bigger than class `y`, but in reality, they have no numerical correlation.

Another way to go, is to transform each feature cloumn into several features columns, with information about which class is the real one for each input. This is called [one hot encoding](https://www.quora.com/What-is-one-hot-encoding-and-when-is-it-used-in-data-science). Here is an example of how the one hot encoding will look like for the `cap-surface` column:

![One hot encoding example](notebook_images/one_hot_encoding_example.png)

We will use this technique for each feature of the data

In [27]:
data_matrix = []

for column in data.columns:
    print(column)
    column_data = []
    column_values = list(data[column])
    print(column_values, len(column_values))
    counter = Counter(column_values) #Counter({'x': 68, 'b': 29, 'f': 16, 's': 6})
    print(counter)
    n_column_labels = len(counter)
    print(n_column_labels)
    column_labels = list(counter.keys())
    print(column_labels)
    
    for i in range(n_column_labels):
        column_data.append(np.zeros(len(column_values)))
        print(i,column_data)
        
    for j in range(len(column_values)):
        label_index = column_labels.index(column_values[j])
        column_data[label_index][j] = 1
        
    for data_column in column_data:
        data_matrix.append(data_column)
        
data_matrix = np.array(data_matrix).transpose()
data_matrix

cap-shape
['x', 'x', 'b', 'x', 'x', 'x', 'b', 'b', 'x', 'b', 'x', 'x', 'b', 'x', 'x', 's', 'f', 'x', 'x', 'x', 'b', 'x', 'b', 'b', 'b', 'f', 'x', 'x', 'f', 'x', 'b', 'x', 'x', 'x', 'b', 'x', 's', 'x', 'x', 'b', 'b', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'x', 'f', 'x', 'x', 'b', 'x', 'x', 'b', 'f', 'b', 'x', 'x', 's', 'b', 'b', 'b', 'b', 'f', 'x', 'f', 'x', 'x', 'f', 'b', 'f', 'x', 'b', 'f', 'x', 'f', 'x', 'f', 'x', 'x', 'f', 'x', 'x', 'x', 'b', 'x', 'f', 's', 'x', 'b', 'x', 'x', 'x', 'x', 'f', 'x', 'b', 'x', 'x', 'b', 'f', 'x', 'b', 'x', 'x', 'b', 'b', 'x', 'x', 's', 'x', 'x', 'x', 'x', 's', 'x', 'x', 's', 'x', 'x', 'f', 'f', 'x', 'x', 'b', 'f', 'x', 'b', 'b', 'b', 'f', 'x', 'f', 'x', 'f', 'x', 'x', 'b', 'x', 'b', 's', 'f', 'x', 'x', 'f', 'x', 'b', 'b', 'x', 'x', 'x', 's', 'x', 'x', 'b', 'x', 'b', 'b', 'b', 'b', 'x', 'f', 'x', 'f', 'b', 'b', 'x', 'b', 'x', 'b', 'x', 'b', 'f', 'b', 'b', 'b', 'x', 'b', 'x', 's', 'f', 'x', 'f', 'x', 'b', 'f', 'b', 'x', 'x', 'x', 'b', 'f', 'x', 'f', 'b', 'x',

array([[1., 0., 0., ..., 0., 0., 0.],
       [1., 0., 0., ..., 0., 0., 0.],
       [0., 1., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 0., 0., ..., 0., 0., 1.],
       [1., 0., 0., ..., 0., 0., 1.]])

In [28]:
data_matrix.shape

(8124, 117)

We still have 8124 data points, but now we have 117 features.


## Split the data into training, validation and test sets

Now we split the dataset and labels, into training, validation and test sets. We will use the 10% of the total data for validation, the 10% for testing, and the 80% for training.

Fisrt we calculate how many data points go to each set:

In [29]:
data_split = 0.1
split_instances = round(data_matrix.shape[0] * data_split)
training_instances = data_matrix.shape[0] - split_instances * 2
print("Training instances: {}".format(training_instances))
print("Validation instances: {}".format(split_instances))
print("Test instances: {}".format(split_instances))

Training instances: 6500
Validation instances: 812
Test instances: 812


### Training set

In [30]:
training_data = data_matrix[:training_instances, :]
training_labels = labels[:training_instances, :]
print("Training data shape = {}".format(training_data.shape))

Training data shape = (6500, 117)


### Validation set

In [31]:
validation_data = data_matrix[training_instances:training_instances + split_instances, :]
validation_labels = labels[training_instances:training_instances + split_instances, :]
print("Validation data shape = {}".format(validation_data.shape))

Validation data shape = (812, 117)


### Test set

In [32]:
test_data = data_matrix[training_instances + split_instances:training_instances + split_instances * 2, :]
test_labels = labels[training_instances + split_instances:training_instances + split_instances *2, :]
print("Testing data shape = {}".format(test_data.shape))

Testing data shape = (812, 117)


## Save the processed data to files

We will save the numpy arrays generated to .npy file (numpy files). We will save 6 different files:
- training data
- training labels
- validation data
- validation labels
- test data
- test labels

In [33]:
Path("training_data").mkdir(exist_ok=True)
np.save(open("training_data/mushrooms_training_data.npy", 'wb'), training_data)
np.save(open("training_data/mushrooms_training_labels.npy", 'wb'), training_labels)
np.save(open("training_data/mushrooms_validation_data.npy", 'wb'), validation_data)
np.save(open("training_data/mushrooms_validation_labels.npy", 'wb'), validation_labels)
np.save(open("training_data/mushrooms_test_data.npy", 'wb'), test_data)
np.save(open("training_data/mushrooms_test_labels.npy", 'wb'), test_labels)